packages/jsActions/native-mobile-resources/src/camera/TakePicture.ts

```diff
@@ -50,117 +109,199 @@ export async function TakePicture(
     }
 
-    function getPictureMethod(): (
-        options: ImagePickerOptions,
-        callback: (response: ImagePickerResponse) => void
-    ) => void {
-        const source = pictureSource ? pictureSource : "either";
+    async function getPictureMethod(): Promise<
+        (
+            options: ImagePickerV2Options | CameraOptions | ImageLibraryOptions,
+            callback: (response: ImagePickerV2Response | ImagePickerResponse) => void
+        ) => void
+    > {
+        async function handleCameraRequest(): Promise<typeof launchCamera> {
+            if (Platform.OS === "android" && nativeVersionMajor === 4) {
+                await checkAndMaybeRequestAndroidPermission();
+            }
 
-        switch (source) {
+            return launchCamera;
+        }
+
+        switch (pictureSource) {
             case "imageLibrary":
-                return ImagePicker.launchImageLibrary;
+                return launchImageLibrary;
+
             case "camera":
-                return ImagePicker.launchCamera;
-            case "either":
+                return handleCameraRequest();
+
             default:
-                return ImagePicker.showImagePicker;
+                return handleCameraRequest();
         }
     }
 
-    function getOptions(): ImagePickerOptions {
+    async function checkAndMaybeRequestAndroidPermission(): Promise<void> {
+        let requestResult: string;
+
+        async function requestAndroidPermission(): Promise<string> {
+            requestResult = await RNPermissions.request(RNPermissions.PERMISSIONS.ANDROID.CAMERA);
+
+            if (requestResult === RNPermissions.RESULTS.DENIED) {
+                // re-enter request flow. note, if a request is denied twice, result = blocked.
+                requestResult = await requestAndroidPermission();
+            }
+
+            return requestResult;
+        }
+
+        // https://github.com/zoontek/react-native-permissions#android-flow
+        const statusResult = await RNPermissions.check(RNPermissions.PERMISSIONS.ANDROID.CAMERA);
+
+        if (statusResult === RNPermissions.RESULTS.UNAVAILABLE) {
+            throw new Error("The camera is unavailable.");
+        }
+
+        if (statusResult === RNPermissions.RESULTS.BLOCKED) {
+            throw new Error("Camera access for this app is currently blocked.");
+        }
+
+        if (statusResult === RNPermissions.RESULTS.DENIED) {
+            requestResult = await requestAndroidPermission();
+
+            if (requestResult === RNPermissions.RESULTS.BLOCKED) {
+                throw new Error("Camera access for this app is currently blocked.");
+            }
+        }
+    }
+
+    function getOptions(): ImagePickerV2Options | CameraOptions | ImageLibraryOptions {
         const { maxWidth, maxHeight } = getPictureQuality();
-        const [language] = getLocales().map(local => local.languageCode);
-        const isDutch = language === "nl";
+
+        if (nativeVersionMajor === 2) {
```

Comment:
I think a nicer way to handle this is to define two functions, one for v2 and one for v4, and then inject that function into `getOptions`. You can also consider extracting all version specific functions into their own files.

Reply:
Updated